docs/tutorials/data_provider.md

```diff
@@ -62,6 +62,83 @@ impl AdditiveIdentity {
 }
 ```
 
+## Loading Additional Data at Runtime
+
+A key feature of ICU4X is the ability to download data dynamically. A key use case is to add additional locales at runtime.
+
+Dynamic data loading can currently be performed in userland. A future core library API may provide this functionality; please submit feedback in [#2985](https://github.com/unicode-org/icu4x/issues/2985).
+
+```rust
+use icu_provider_adapters::either::EitherProvider;
+use icu_provider_adapters::fallback::LocaleFallbackProvider;
+use icu_provider_adapters::fork::ForkByKeyProvider;
+use icu_provider_adapters::fork::MultiForkByErrorProvider;
+use icu_provider_adapters::fork::predicates::MissingLocalePredicate;
+use icu_provider_blob::BlobDataProvider;
+use icu_provider_fs::FsDataProvider;
+use icu_provider::DataLocale;
+use icu_provider::hello_world::HelloWorldFormatter;
+use icu::locid::locale;
+use icu::locid::subtags::Language;
+use std::path::PathBuf;
+use writeable::Writeable;
+
+// Our growable data provider will be a fallback-enabled ForkByKeyProvider forking between a
+// BlobDataProvider with static core data and a MultiForkByErrorProvider with runtime-loaded data.
+type GrowableDataProvider = LocaleFallbackProvider<
```

Comment:
I'd rather see a `GrowableDataProvider` struct that implements `load_buffer` than to have the closure that also contains the constructor. That'd be a way more general solution and shouldn't be more complex.

You can also take out the `LocaleFallbackProvider`, it doesn't contribute to the logic and should be able to be wrapped around the `GrowableDataProvider` as needed (with an explicit `LocaleFallbacker` loaded from either postcard or baked). In fact, `de-CH` currently only loads `de` because you explicitly load `locale.language`. If you wrapped `LocaleFallbackProvider` around this, it will try loading the actual fallback chain, so `gsw` will load correctly. This implementation works for `gws` on the second try only (well, if it didn't panic).

Reply:
So apparently `gws` doesn't fall back to `de-CH`. Shouldn't there be a `gws` -> `gws-CH` -> `und-CH` -> `de-CH` chain or something?

Reply:
https://github.com/unicode-org/icu4x/blob/069ee4254b7081924e335a5ca31a68dd3243c8f0/docs/tutorials/data_provider.md#loading-additional-data-at-runtime

Reply:
It was a conscious decision to not make a struct implementing BufferProvider because

1. It's much easier to get wrong
2. Your example is basically a thread-safe MultiForkByErrorProvider; if we need such a thing, it should go into the core library; but I don't want to do that yet because we don't have enough real-world examples
3. The code is more complicated for the reader; it's a _tutorial_

I'm not opposed to the idea of making GrowableDataProvider a struct, with a public field you can access. I just think that it shouldn't involve implementing the traits in user land. There should be a whole book chapter about how to correctly implement the traits. This isn't the right place for that.

Reply:
> In fact, de-CH currently only loads de because you explicitly load locale.language.

The actual data loading uses the full locale. I pull out locale.language only during loading of the language pack. The language pack could contain multiple regional variants.

Reply:
> In fact, de-CH currently only loads de because you explicitly load locale.language.

The actual data loading uses the full locale. I pull out locale.language only during loading of the language pack. The language pack could contain multiple regional variants.

Reply:
Can you take out the locale fallback? It distracts from the thing you're trying to demonstrate, and doing any resolution with `locale.language` is not something we should put in users' heads.

Reply:
I assumed that people probably want locale fallback enabled. But I can fully disable it and then just load the one helloworld key for the two languages.

Reply:
Sure, but they can figure out how to add a LocaleFallbackProvider where needed. Let's keep example minimal.